src/mpid/ch4/src/ch4r_rma_target_callbacks.c

```diff
@@ -1445,24 +1450,18 @@ int MPIDIG_put_iov_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_A
 
     offset = win->disp_unit * msg_hdr->target_disp;
     base = MPIDIG_win_base_at_target(win);
+    MPIDIG_REQUEST(rreq, buffer) = (void *) (offset + base);
+    MPIDIG_REQUEST(rreq, count) = msg_hdr->data_sz;
 
-    MPIDIG_REQUEST(rreq, req->target_cmpl_cb) = put_iov_target_cmpl_cb;
+    MPIDIG_REQUEST(rreq, req->target_cmpl_cb) = put_dt_target_cmpl_cb;
 #ifndef MPIDI_CH4_DIRECT_NETMOD
     MPIDI_REQUEST(rreq, is_local) = is_local;
 #endif
 
-    /* Base adjustment for iov will be done after we get the entire iovs,
-     * at MPIDIG_put_data_target_msg_cb */
-    MPIR_Assert(msg_hdr->n_iov);
-    dt_iov = (struct iovec *) MPL_malloc(sizeof(struct iovec) * msg_hdr->n_iov, MPL_MEM_RMA);
-    MPIR_Assert(dt_iov);
-
-    MPIDIG_REQUEST(rreq, req->preq.dt_iov) = dt_iov;
-    MPIDIG_REQUEST(rreq, req->preq.n_iov) = msg_hdr->n_iov;
-    MPIDIG_REQUEST(rreq, req->preq.target_addr) = (void *) (offset + base);
-
-    int data_sz = msg_hdr->n_iov * sizeof(struct iovec);
-    MPIDIG_recv_init(1, in_data_sz, dt_iov, data_sz, rreq);
+    MPIR_Assert(msg_hdr->flattened_sz);
+    void *flattened_dt = MPL_malloc(msg_hdr->flattened_sz, MPL_MEM_BUFFER);
```

Comment:
Can this flattened type be cached inside the `MPIR_Datatype` handle, instead of creating it every time?  Datatype creation is infrequent.  Flattening the datatype every time seems inefficient.

Reply:
Makes sense. I will make that change and report shortly.

Reply:
Actually, on second thought, that could perhaps be done in a separate PR, given that this PR is time critical.

Reply:
Either way is fine with me. Happy to merge this as-is.